File: 155_MinStack/155.py
class MinStack(object):
    
    def __init__(self):
        """
        initialize your data structure here.
        """
        self.stack = []
        self.minVal = sys.maxint

    def push(self, x):
        """
        :type x: int
        :rtype: void
        """
        if x < self.minVal:
            self.minVal = x
        self.stack.append(x)

    def pop(self):
        """
        :rtype: void
        """
        self.stack.pop()
        if len(self.stack) is not 0:
            self.minVal = min(self.stack)
        else:
            self.minVal = sys.maxint
        

    def top(self):
        """
        :rtype: int
        """
        return self.stack[-1]

    def getMin(self):
        """
        :rtype: int
        """
        return self.minVal


# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(x)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()

File: 155_MinStack/155.java
public class MinStack {

    Stack<Integer> stack;
    int min;

    /** initialize your data structure here. */
    public MinStack() {
        this.stack = new Stack<Integer>();
        this.min = Integer.MAX_VALUE;
    }
    
    public void push(int x) {
        if (x <= this.min){
            this.stack.push(min);
            this.min = x;
        }
        this.stack.push(x);
    }
    
    public void pop() {
        if (this.stack.pop() == this.min){
            this.min = this.stack.pop();
        }
    }
    
    public int top() {
        return this.stack.peek();
    }
    
    public int getMin() {
        return this.min;
    }
}

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack obj = new MinStack();
 * obj.push(x);
 * obj.pop();
 * int param_3 = obj.top();
 * int param_4 = obj.getMin();
 */